apps/simple-caching/index.js

```diff
@@ -0,0 +1,31 @@
+import { parse, resolve } from 'url'
+
+/*
+  This is the site we're going to proxy. You can put it in your `.fly.yml` file and/or do `fly secret set HOST https://my-uncached-website.com` in the command line after installing `@fly/fly`
+*/
+const destination = app.config.HOST
+
+fly.http.respondWith(async (req) => {
+  /* This just takes apart and reconstructs the url. Not super useful in this example, but I mean to use the params object later to do some other fly.io magic. Also protects against stuff like HOST having or not having a backslash. */
+  const params = parse(req.url)
+  const path = resolve(destination, params.path)
+  /* Caches are stored as keys and values, that simple. You can store `fly.cache.getString()` accepts a key and returns a string version of whatever is stored. The first time a user accesses this Edge App, cached is going to be null or undefined, cause we've never stored anything with any key. */
+  let cached = await fly.cache.getString(path)
+  /* No cache? Lets do something about that. */
+  if (!cached) {
+    /* Fetch the requested site. Remember, this just happens if nothing is cached. */
+    const request = await fetch(path, {
+      headers: {
+        Accept: "application/json" // this header makes sure we don't receive .js requests as html, but doesn't seem to bother the html requests
```

Comment:
You can just make a new request from the existing one:

```javascript
const request = new Request(req)
request.url = path // new url, otherwise use existing request
```

Reply:
While I think this did work when doing `fetch(request)` where `request` is the instance of `Request` with `.url = path` like you described, this is apparently a read-only property (according to [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/Request)). I think there is a `Request.clone()` but iirc when I tried to use it in fly it said it wasn't a function? Any, playing around with this and thanks for mentioning it.

Reply:
Think I've worked this out. If you do `const url = new URL(req.url)`, must also do `request.url = url.href`, I think.

Reply:
Ah, derp, I gave you a bad example.

The "right" way to make a new request with a different url is:
 
```javascript
const url = new URL("https://fly.io")
const req2 = new Request(url.toString(), req) // uses new url and all other params from req
```

You are right that the spec says `.url` is read only, we cheated on that one because it's so convenient to change URL.

`req.clone()` should also work but usually you'd only use that if you wanted to read a POST body more than once, not to make a request to a different URL.